File: sdk/go/helpers/pop.go
package helpers

import (
	"context"
	"crypto/ed25519"
	"encoding/base64"
	"errors"
	"fmt"
	"net/http"
	"strings"
)

// Agent-binding proof of possession for signed delivery URLs (ADR-013).
//
// When a signed URL carries an agent_id — the RFC 7638 thumbprint of the key
// that signed the offer acceptance — a code-capable edge requires the fetcher to
// prove possession of that key, fully offline. The proof is two headers on the
// GET: the raw public key in X-RAMP-Agent-Key, and an RFC 9421 signature over
// @method + @target-uri. The edge then enforces a three-way identity:
//
//	agent_id (URL) == keyid (Signature-Input) == thumbprint(presented key)
//
// The last equality is the one that cannot be dropped. Verifying the signature
// against the presented key alone proves nothing — any actor could present their
// own key plus a valid self-signature.
//
// This is the SIGN face. The verify face ships in sdk/ts and sdk/python (the
// edge runs there); Go is the byte oracle both are pinned to, through
// testdata/pop-vectors.json.

// AgentKeyHeader carries the raw Ed25519 public key the fetcher presents, as
// base64url with no padding. ADR-013 chose a dedicated header over an inline JWK
// in keyid: the edge hashes this value and requires the digest to equal the
// URL's agent_id, so a fetcher cannot present one key while naming another.
const AgentKeyHeader = "X-RAMP-Agent-Key"

// popLabel is the only signature label this profile emits. Unlike the RAMP
// request profile there is no forwarding chain here — a delivery fetch is a
// single hop to the edge — so sigN>1 never arises and the label is fixed rather
// than computed.
const popLabel = "sig1"

// ErrMissingTargetURI signals a proof requested without the URL it is meant to
// bind. @target-uri is half the covered set; signing without it would produce a
// proof valid for any URL the presented key is offered against, which is the
// replay this profile exists to stop.
var ErrMissingTargetURI = errors.New("helpers: missing target URI (required by the agent-binding profile)")

// ErrKeyIDMismatch signals that the keyid does not match the RFC 7638
// thumbprint of the key presented alongside it. The edge checks the same
// equality and answers thumbprint_mismatch, but by then the cause — a custody
// layer that paired a keyid with the wrong key — is several hops from the
// symptom. Refusing here names it at the source.
var ErrKeyIDMismatch = errors.New("helpers: keyid is not the thumbprint of the presented key")

// ErrInvalidPoPInput signals a proof input that cannot be written into a
// signature base without changing its shape — a control byte in the method or the
// target URI, which the line-delimited base would read as a component boundary.
var ErrInvalidPoPInput = errors.New("helpers: proof input is not usable in a signature base")

// isControlByte reports whether r is a C0 control or DEL. Applied to the two
// values written verbatim into the signature base.
func isControlByte(r rune) bool { return r < 0x20 || r == 0x7f }

// PoPOptions carries what a delivery-URL proof of possession needs beyond the
// key material. Only URL, Created and Expires are required.
type PoPOptions struct {
	// URL is the signed delivery URL, used VERBATIM as @target-uri — the exact
	// bytes the Exchange minted, query parameters and all.
	//
	// This is a string and not a request value on purpose. The edge rebuilds the
	// base from the raw request line it received, so the signing side must not
	// route the URL through a parsed value first: doing so yields the DECODED
	// path, expanding every percent-escape before it reaches the signed bytes.
	// %2F is the sharpest case, since it decodes to a real separator and the
	// signature would then cover a different path structure than the wire carried.
	// The result is a proof that cannot verify, surfacing as a blanket 403 with no
	// indication that the URL was the problem.
	URL string
	// KeyID is the RFC 9421 keyid: the RFC 7638 thumbprint of the presented key.
	// It is the anchor of the three-way identity the edge enforces. Empty means
	// "take the Signer's own key id". Either way it is cross-checked against the
	// presented key before anything is signed.
	KeyID string
	// Created is the unix-seconds instant the proof was made. The edge rejects a
	// created more than 300s in its own future, so a signer whose clock runs fast
	// fails closed.
	Created int64
	// Expires is the unix-seconds cutoff after which the proof is stale. Keep it
	// short: the covered set is only method and URL, so within this window the
	// proof is replayable by anyone who observes the request.
	Expires int64
	// Method is the HTTP method being signed. Empty means GET. A signed URL is
	// read-only in practice, but @method is covered precisely so a proof made for
	// a GET cannot be lifted onto a write.
	Method string
}

// AgentBinding is the proof a fetcher attaches to a bound delivery request: the
// three header values, ready to apply. It is returned as values rather than
// written onto a request so a caller can sign before it has built one, so this
// tier stays free of any dialing surface, and so the emitted bytes can be
// asserted directly against the shared cross-language vectors.
type AgentBinding struct {
	// AgentKey is the X-RAMP-Agent-Key value: base64url, no padding.
	AgentKey string
	// SignatureInput is the full Signature-Input value, label included.
	SignatureInput string
	// Signature is the full Signature value, label included. The byte string
	// inside the colons is STANDARD base64 while AgentKey above is base64url — an
	// asymmetry that comes from RFC 8941's byte-sequence encoding meeting a header
	// this profile defines itself, and one a verifier will not forgive.
	Signature string
}

// Apply writes the binding's three headers onto h.
func (b AgentBinding) Apply(h http.Header) {
	h.Set(AgentKeyHeader, b.AgentKey)
	h.Set("Signature-Input", b.SignatureInput)
	h.Set("Signature", b.Signature)
}

// popSignatureBase builds the RFC 9421 signature base for the agent-binding
// profile: the two covered components followed by the parameters line, joined
// with newlines and with no trailing newline.
//
// It takes raw strings rather than a request value for the reason PoPOptions.URL
// documents — the verbatim URL is the contract — and because the RAMP base
// builder in sigbase.go reconstructs @target-uri from a parsed URL's decoded
// path, which is exactly the transformation this profile must not perform. The
// two builders are pinned against each other in the internal base test: they
// agree on every ordinary delivery URL and diverge only where the path carries a
// percent-escape.
//
// rawParams is taken as given rather than rebuilt: on the verifying side it
// arrives verbatim in the Signature-Input header, so treating it as an opaque
// string here is what keeps the two faces symmetric.
func popSignatureBase(method, rawURL, rawParams string) string {
	return strings.Join([]string{
		`"@method": ` + strings.ToUpper(method),
		`"@target-uri": ` + rawURL,
		`"@signature-params": ` + rawParams,
	}, "\n")
}

// popSignatureParams renders the @signature-params value for this profile.
//
// The parameter ORDER is part of the wire contract, not a formatting choice: the
// base is signed over this exact string and the verifier reconstructs it from the
// header as received. keyid, alg, created, expires — matching the TypeScript and
// Python faces. A generic RFC 9421 emitter tends to produce
// created;expires;alg;keyid instead, and that mismatch is the whole reason this
// profile builds its own base instead of routing through SignRequest.
func popSignatureParams(keyID string, created, expires int64) string {
	return fmt.Sprintf(
		`("@method" "@target-uri");keyid=%q;alg=%q;created=%d;expires=%d`,
		keyID, AlgEd25519, created, expires,
	)
}

// SignAgentBinding produces the proof of possession a fetcher presents when it
// retrieves a delivery URL bound to an agent key. The covered set is exactly
// @method and @target-uri: a GET carries no body to digest, and the signed URL is
// itself the credential, already covered by @target-uri, so there is no
// Authorization header worth binding. That is why SignRequest cannot serve this
// profile — it enforces the five-component RAMP set.
//
// The key arrives as a Signer plus the public half rather than as a raw private
// key: custody stays with the application (a KMS or HSM signer never exposes its
// key), and the public half must be supplied separately because the presented-key
// header carries it and a Signer cannot yield it.
func SignAgentBinding(ctx context.Context, signer Signer, pub ed25519.PublicKey, opts PoPOptions) (AgentBinding, error) {
	keyID, err := validateAgentBinding(signer, pub, opts)
	if err != nil {
		return AgentBinding{}, err
	}
	method := opts.Method
	if method == "" {
		method = http.MethodGet
	}
	params := popSignatureParams(keyID, opts.Created, opts.Expires)
	raw, err := signer.Sign(ctx, []byte(popSignatureBase(method, opts.URL, params)))
	if err != nil {
		return AgentBinding{}, fmt.Errorf("helpers: sign agent binding: %w", err)
	}
	return AgentBinding{
		AgentKey:       base64.RawURLEncoding.EncodeToString(pub),
		SignatureInput: popLabel + "=" + params,
		Signature:      popLabel + "=:" + base64.StdEncoding.EncodeToString(raw) + ":",
	}, nil
}

// validateAgentBinding checks every precondition of a bound fetch and returns the
// keyid the proof will declare. It runs before any signing so a caller that has
// mispaired a key and a keyid learns it here rather than from a 403 that names
// nothing.
func validateAgentBinding(signer Signer, pub ed25519.PublicKey, opts PoPOptions) (string, error) {
	if signer == nil {
		return "", errors.New("helpers: agent-binding signer is nil")
	}
	if len(pub) != ed25519.PublicKeySize {
		return "", fmt.Errorf("helpers: ed25519 public key must be %d bytes, got %d", ed25519.PublicKeySize, len(pub))
	}
	if opts.URL == "" {
		return "", ErrMissingTargetURI
	}
	// The base is line-delimited and both values are written into it verbatim, so
	// a control byte in either would add or split a line and the bytes signed here
	// would stop describing the request the verifier reconstructs. Refused rather
	// than escaped: no legitimate method or target URI contains one, and a
	// signature base is the wrong place to be lenient.
	if i := strings.IndexFunc(opts.URL, isControlByte); i >= 0 {
		return "", fmt.Errorf("%w: target URI carries a control byte at %d", ErrInvalidPoPInput, i)
	}
	if i := strings.IndexFunc(opts.Method, isControlByte); i >= 0 {
		return "", fmt.Errorf("%w: method carries a control byte at %d", ErrInvalidPoPInput, i)
	}
	// A proof carrying no created sails through as a signature claiming 1970: the
	// edge bounds how far created may lead its clock, not how far it may lag, so
	// freshness would be silently absent rather than loudly wrong.
	if opts.Created <= 0 {
		return "", ErrMissingCreated
	}
	if opts.Expires <= 0 {
		return "", ErrMissingExpires
	}
	if signer.Algorithm() != AlgEd25519 {
		return "", fmt.Errorf("%w: agent binding requires %q, signer offers %q",
			ErrUnsupportedAlgorithm, AlgEd25519, signer.Algorithm())
	}
	keyID := opts.KeyID
	if keyID == "" {
		keyID = signer.KeyID()
	}
	thumb, err := Thumbprint(pub)
	if err != nil {
		return "", fmt.Errorf("helpers: derive keyid thumbprint: %w", err)
	}
	if keyID != thumb {
		return "", fmt.Errorf("%w: keyid %q, presented-key thumbprint %q", ErrKeyIDMismatch, keyID, thumb)
	}
	return keyID, nil
}

File: sdk/python/ramp_sdk/pop.py
"""RFC 9421 GET proof-of-possession verification (ADR-013) — pure L1 helper.

Mirrors the sdk/ts sibling (sdk/ts/src/pop.ts ``verifyAgentBinding``). When a
signed URL carries an ``agent_id`` (the agent's RFC 7638 thumbprint), a
code-capable edge requires the fetcher to prove possession of the bound key,
fully offline:

  1. present its raw Ed25519 public key in ``X-RAMP-Agent-Key``, and
  2. sign the GET with RFC 9421 over ``@method`` + ``@target-uri``,

and the edge enforces the 3-way identity:

  agent_id (URL) == keyid (Signature-Input) == thumbprint(presented key)

The Ed25519 verify primitive is INJECTABLE (``verify_ed25519``) so a runtime
without a preferred Ed25519 backend can supply its own — the byte contract is the
signature base, not the primitive. The default primitive uses ``cryptography``.
``now`` is INJECTED (unix seconds) so the verify reads no wall clock. Byte-parity
guard: the pop vectors are produced by the sdk/go signer over the same
@method/@target-uri base (sigbase.go buildSignatureBase).
"""

from __future__ import annotations

import base64
import re
from collections.abc import Callable, Mapping
from dataclasses import dataclass

from cryptography.exceptions import InvalidSignature
from cryptography.hazmat.primitives.asymmetric.ed25519 import (
    Ed25519PrivateKey,
    Ed25519PublicKey,
)

from .b64 import b64url_decode, b64url_nopad
from .thumbprint import thumbprint

AGENT_KEY_HEADER = "x-ramp-agent-key"

_ED25519_PUBLIC_KEY_BYTES = 32
#: The GET PoP covers exactly two components: @method and @target-uri.
_POP_COVERED_COUNT = 2
#: A proof's created timestamp may not lead the verifier clock by more than this.
_MAX_FUTURE_SKEW_SEC = 300

#: Injected Ed25519 verify primitive: ``(pubkey, signature, message) -> valid?``.
Ed25519Verify = Callable[[bytes, bytes, bytes], bool]


@dataclass(frozen=True)
class PopResult:
    """Verdict of a proof-of-possession verification."""

    ok: bool
    reason: str | None = None


@dataclass(frozen=True)
class _SignatureInput:
    covered: list[str]
    raw_params: str
    keyid: str
    alg: str
    created: int | None
    expires: int | None


def _default_verify_ed25519(pubkey: bytes, signature: bytes, message: bytes) -> bool:
    try:
        Ed25519PublicKey.from_public_bytes(pubkey).verify(signature, message)
    except (InvalidSignature, ValueError):
        return False
    return True


def _read_presented_key(headers: Mapping[str, str]) -> bytes | None:
    raw = headers.get(AGENT_KEY_HEADER)
    if not raw:
        return None
    try:
        key = b64url_decode(raw)
    except (ValueError, TypeError):
        return None
    if len(key) != _ED25519_PUBLIC_KEY_BYTES:
        return None
    return key


def _parse_signature_input(raw: str | None) -> _SignatureInput | None:
    """Parse ``label=("@method" "@target-uri");keyid="..";alg="..";created=..;expires=..``."""
    if not raw:
        return None
    eq = raw.find("=")
    if eq < 0:
        return None
    raw_params = raw[eq + 1 :].strip()
    m = re.match(r"^\(([^)]*)\)(.*)$", raw_params)
    if not m:
        return None
    covered = re.findall(r'"([^"]+)"', m.group(1))
    tail = m.group(2)
    keyid = _match_param(tail, r';keyid="([^"]*)"')
    alg = _match_param(tail, r';alg="([^"]*)"')
    if keyid is None or alg is None:
        return None
    return _SignatureInput(
        covered=covered,
        raw_params=raw_params,
        keyid=keyid,
        alg=alg,
        created=_match_int(tail, r";created=(\d+)"),
        expires=_match_int(tail, r";expires=(\d+)"),
    )


def _match_param(s: str, pattern: str) -> str | None:
    m = re.search(pattern, s)
    return m.group(1) if m else None


def _match_int(s: str, pattern: str) -> int | None:
    m = re.search(pattern, s)
    return int(m.group(1)) if m else None


def _parse_signature(raw: str | None) -> bytes | None:
    """Parse the RFC 9421 ``Signature`` header value ``label=:<base64>:``."""
    if not raw:
        return None
    m = re.search(r":([^:]+):", raw)
    if not m:
        return None
    try:
        # The Signature byte string is STANDARD base64 (not b64url-nopad).
        return base64.b64decode(m.group(1))
    except (ValueError, TypeError):
        return None


def _covers_exactly(covered: list[str]) -> bool:
    return len(covered) == _POP_COVERED_COUNT and "@method" in covered and "@target-uri" in covered


def signature_base(method: str, url: str, raw_params: str) -> str:
    """Rebuild the RFC 9421 signature base for the covered GET components.

    MUST stay byte-identical to the sdk/go signer (sigbase.go buildSignatureBase).
    """
    return "\n".join(
        [
            f'"@method": {method.upper()}',
            f'"@target-uri": {url}',
            f'"@signature-params": {raw_params}',
        ]
    )


def sign_agent_binding(
    *,
    url: str,
    signer_seed: bytes,
    created: int,
    expires: int,
) -> tuple[str, str, str]:
    """Produce the agent's GET proof-of-possession — the SIGN face of this module.

    Returns ``(presented_key_b64url, signature_input_value, signature_value)``:
    the ``X-RAMP-Agent-Key`` value, the ``Signature-Input`` header value, and the
    ``Signature`` header value the fetcher attaches to its GET. The covered set is
    exactly ``@method @target-uri`` (ADR-013), keyid is the RFC 7638 thumbprint of
    the signer's public key (the 3-way identity anchor), and the signed bytes are
    the same :func:`signature_base` the verify face reconstructs — byte-identical
    to the sdk/go signer (pinned by pop-vectors.json). ``created``/``expires`` are
    injected unix seconds: the helper reads no clock (L1-pure).
    """
    # The signature base is line-delimited and ``url`` is written into it verbatim,
    # so a control byte would add or split a component line and the bytes signed
    # here would stop describing the request a verifier reconstructs. Refused
    # rather than escaped: no legitimate target URI contains one, and a signature
    # base is the wrong place to be lenient. Mirrors the Go signer, which refuses
    # the same bytes in both the method and the URL; the method arm does not apply
    # here because this face always signs GET.
    #
    # Scanned over the UTF-8 BYTES, not the code points, so the reported offset is
    # the same number Go's strings.IndexFunc reports for the same input. Which
    # inputs are refused is unaffected — a control byte is always a single UTF-8
    # byte — but an unlabelled index under identical wording meant three units.
    raw = url.encode("utf-8")
    bad = next((i for i, b in enumerate(raw) if b < 0x20 or b == 0x7F), None)
    if bad is not None:
        raise ValueError(f"target URI carries a control byte at byte {bad}")

    priv = Ed25519PrivateKey.from_private_bytes(signer_seed)
    pub = priv.public_key().public_bytes_raw()
    keyid = thumbprint(pub)
    sig_params = (
        f'("@method" "@target-uri");keyid="{keyid}";alg="ed25519"'
        f";created={created};expires={expires}"
    )
    raw = priv.sign(signature_base("GET", url, sig_params).encode())
    # The Signature byte string is STANDARD base64 (mirror _parse_signature).
    return (
        b64url_nopad(pub),
        f"sig1={sig_params}",
        f"sig1=:{base64.b64encode(raw).decode()}:",
    )


def _freshness_failure(parsed: _SignatureInput, now: int) -> str | None:
    if parsed.created is None:
        return "pop_missing_created"
    if parsed.created > now + _MAX_FUTURE_SKEW_SEC:
        return "pop_future_created"
    if parsed.expires is None:
        return "pop_missing_exp"
    if now >= parsed.expires:
        return "pop_expired"
    return None


def verify_agent_binding(
    *,
    method: str,
    url: str,
    headers: Mapping[str, str],
    agent_id: str,
    now: int,
    verify_ed25519: Ed25519Verify | None = None,
) -> PopResult:
    """Verify the agent's proof of possession of the key bound to ``agent_id``.

    Returns ``ok=True`` only when the presented key, the RFC 9421 signature, and
    the 3-way identity all check out. ``now`` is unix seconds (injected).
    """
    presented = _read_presented_key(headers)
    if presented is None:
        return PopResult(ok=False, reason="missing_agent_key")

    parsed = _parse_signature_input(headers.get("signature-input"))
    if parsed is None:
        return PopResult(ok=False, reason="malformed_sig_input")
    if parsed.alg.lower() != "ed25519":
        return PopResult(ok=False, reason="unsupported_alg")
    if not _covers_exactly(parsed.covered):
        return PopResult(ok=False, reason="bad_covered_components")

    sig_bytes = _parse_signature(headers.get("signature"))
    if sig_bytes is None:
        return PopResult(ok=False, reason="missing_sig")

    # 3-way identity: keyid and the presented-key thumbprint must both equal the
    # URL-bound agent_id before any signature work is trusted.
    if parsed.keyid != agent_id:
        return PopResult(ok=False, reason="keyid_mismatch")
    if thumbprint(presented) != agent_id:
        return PopResult(ok=False, reason="thumbprint_mismatch")

    stale = _freshness_failure(parsed, now)
    if stale is not None:
        return PopResult(ok=False, reason=stale)

    base = signature_base(method, url, parsed.raw_params).encode()
    verify = verify_ed25519 if verify_ed25519 is not None else _default_verify_ed25519
    if not verify(presented, sig_bytes, base):
        return PopResult(ok=False, reason="pop_sig_invalid")
    return PopResult(ok=True)
